server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java

```diff
@@ -569,8 +584,12 @@ private boolean checkForChangedProducts(Product incomingProduct, Set<Product> in
             !currentProvided.equals(incomingProvided);
 
         // Check if the existing product is in the set of changed products
-        if (!productsChanged && changedProducts != null && pid != null) {
-            productsChanged = (changedProducts.get(pid) != null);
+        if (!productsChanged && changedProducts != null) {
+            if (pid != null) {
+                productsChanged = (changedProducts.get(pid) != null);
+            }
+            productsChanged = productsChanged ||
```

Comment:
We can condense this whole block down to a (wrapped?) one-liner:

```
productsChanged |= (pid != null && changedProducts.get(pid) != null) || changedProductsInSet(...);
```

If you opt not to change it, then you need to fix the bug on line 589 which could revert the previous state of productsChanged if the declaration is true on 581 and the pid check is false on 589.

Reply:
are you sure there is a bug?